ddtrace/contrib/botocore/patch.py

```diff
@@ -52,23 +69,155 @@ def inject_trace_data_to_message_attributes(trace_data, entry):
+    Max size per event is 256KB (https://docs.aws.amazon.com/eventbridge/latest/userguide/eb-putevent-size.html)
+    """
+    params = args[1]
+    if "Entries" not in params:
+        log.debug("Unable to inject context. The Event Bridge event had no Entries.")
+        return
+
+    for entry in params["Entries"]:
+        detail = {}
+        if "Detail" in entry:
+            try:
+                detail = json.loads(entry["Detail"])
+            except ValueError:
+                log.debug("Detail is not a valid JSON string")
+                continue
+
+        detail["_datadog"] = {}
+        HTTPPropagator.inject(span.context, detail["_datadog"])
+        detail_json = json.dumps(detail)
+
+        # check if detail size will exceed max size with headers
+        detail_size = len(detail_json)
+        if detail_size >= MAX_EVENTBRIDGE_DETAIL_SIZE:
+            log.debug("Detail with trace injection exceeds max allowed size")
+            continue
+
+        entry["Detail"] = detail_json
+
+
+def get_kinesis_data_object(data):
+    # type: (str, Optional[bool]) -> Dict[str: Any]
+    """
+    :data: the data from a kinesis stream
+    :try_b64: whether we should try to decode the string as base64
+
+    The data from a kinesis stream comes as a string (could be json, base64 encoded, etc.)
+    We support injecting our trace context in the following two cases:
+    - json string
+    - base64 encoded json string
+    If it's neither of these, then we leave the message as it is.
+    """
+
+    # check if data is a json string
+    try:
+        return json.loads(data)
+    except ValueError:
+        pass
+
+    # check if data is a base64 encoded json string
+    try:
+        return json.loads(base64.b64decode(data).decode("ascii"))
+    except ValueError:
+        return None
+
+
+def inject_trace_to_kinesis_stream_data(record, span):
+    # type: (Dict[str, Any], Span) -> None
+    """
+    :record: contains args for the current botocore action, Kinesis record is at index 1
+    :span: the span which provides the trace context to be propagated
+
+    Inject trace headers into the Kinesis record's Data field in addition to the existing
+    data. Only possible if the existing data is JSON string or base64 encoded JSON string
+    Max data size per record is 1MB (https://aws.amazon.com/kinesis/data-streams/faqs/)
+    """
+    if "Data" not in record:
+        log.debug("Unable to inject context. The kinesis stream has no data")
+        return
+
+    data = record["Data"]
+    data_obj = get_kinesis_data_object(data)
+    if data_obj is not None:
+        data_obj["_datadog"] = {}
+        HTTPPropagator.inject(span.context, data_obj["_datadog"])
+        data_json = json.dumps(data_obj)
+
+        # check if data size will exceed max size with headers
+        data_size = len(data_json)
+        if data_size >= MAX_KINESIS_DATA_SIZE:
+            return None
```

Comment:
should raise here else this silently fails? or at least log but I think I'd prefer raising at this point and then it'll be caught by the exception handling